hypothesis/lms
# Unify (a bit) the course and grouping services

For: 

 * https://github.com/hypothesis/lms/issues/3990

Originally I was going to try and get this to pass a mapper along or something, but honestly I think it's too much fuss. There's still quite a bit going on without that:

 * `generate_authority_provided_id` -> `get_authority_provided_id`
    * No longer requires `tool_consumer_instance_guid`
    * As `parent` is optional, it's now the last param and defaults to `None`
 * `upsert_with_parent` -> `upsert_groupings`
    * No longer requires a parent (can be used for courses)
    * Sets `settings` on create
 * `upsert` -> `upsert_course`
    * Uses `upsert_groupings` internally instead of re-implementing

Other changes:

 * The course service is generally `context_id` centric now, rather than `authority_provided_id` centric
 * More test coverage than before

### Testing notes

 * Start all the things
 * Do `make sql`
 * `truncate grouping, grouping_membership;`
 * Open any assignment (e.g. https://hypothesis.instructure.com/courses/125/assignments/873)
 * `select * from grouping` - You should see rows
 * `update grouping set lms_name='SOMEVALUE'`
 * Reload the assignment
 * `select * from grouping` - The name should have been updated
